lib/extractor/layer.ts

```diff
@@ -0,0 +1,141 @@
+import { createReadStream } from "fs";
+import * as minimatch from "minimatch";
+import { basename } from "path";
+import { Readable } from "stream";
+import { extract, Extract } from "tar-stream";
+import { streamToString } from "../stream-utils";
+import { applyCallbacks } from "./callbacks";
+import { ExtractAction, ExtractedLayers, FileNameAndContent } from "./types";
+
+/**
+ * Retrieve the products of files content from the specified docker-archive.
+ * @param dockerArchiveFilesystemPath Path to image file saved in docker-archive format.
+ * @param extractActions Array of pattern-callbacks pairs.
+ * @returns Array of extracted files products sorted by the reverse order of the layers from last to first.
+ */
+export async function extractDockerArchiveLayers(
+  dockerArchiveFilesystemPath: string,
+  extractActions: ExtractAction[],
+): Promise<ExtractedLayers[]> {
+  return new Promise((resolve, reject) => {
+    const tarExtractor: Extract = extract();
+    const layers: Record<string, ExtractedLayers> = {};
+    let layersNames: string[];
+
+    tarExtractor.on("entry", async (header, stream, next) => {
+      if (header.type === "file") {
+        if (isTarFile(header.name)) {
+          layers[header.name] = await extractImageLayer(stream, extractActions);
+        } else if (isManifestFile(header.name)) {
+          layersNames = await getLayersNames(stream);
+        }
+      }
+
+      stream.resume(); // auto drain the stream
+      next(); // ready for next entry
+    });
+
+    tarExtractor.on("finish", () => {
+      resolve(getLayersContent(layersNames, layers));
+    });
+
+    tarExtractor.on("error", reject);
```

Comment:
I guess this one is on me 😬
```suggestion
    tarExtractor.on("error", (error) => reject(error));
```